Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -194,29 +316,25 @@ - (void)URLProtocol:(id)urlProtocol didFailWithError:(id)error {
         [_delegate connection:self didFailWithError:error];
     }
 
-    if (_didRetain && !_didRelease) {
-        _didRelease = TRUE;
-        [self autorelease];
-    }
-    [_delegate autorelease];
+    _protocol = nil;
     _delegate = nil;
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocol:(id)urlProtocol didReceiveResponse:(id)response cacheStoragePolicy:(NSURLCacheStoragePolicy)policy {
+- (void)URLProtocol:(id)urlProtocol didReceiveResponse:(NSURLResponse*)response cacheStoragePolicy:(NSURLCacheStoragePolicy)policy {
     /*
     if ( [response respondsToSelector:@selector(statusCode)] && [response statusCode] != 200 ) {
     [_delegate setError:[NSError errorWithDomain:@"Bad response code" code:[response statusCode] userInfo:nil]];
     }
     */
 
-    _response = [response retain];
+    _response = response;
     _storagePolicy = policy;
 
     if ([_delegate respondsToSelector:@selector(connection:willCacheResponse:)]) {
-        [_delegate connection:self willCacheResponse:response];
+        [_delegate connection:self willCacheResponse:static_cast<NSCachedURLResponse*>(response)];
```

Comment:
whoa, this is scary. I realize it is what was already happening, but those two things are very much not the same thing